### PR TITLE
Debug Mode: Refactor debug event saving

### PR DESCRIPTION
**Pull request by gaetanmaisse**
**Issue**

https://github.com/gravitee-io/issues/issues/6949

Related to PRs:
 - https://github.com/gravitee-io/gravitee-definition/pull/127

**Description**

Introduce a Processor to save the debug event with all the debug steps.

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-xtenvmrvlk.chromatic.com)

🚀 CI was able to deploy the build of this PR, so you can now try it directly [here](https://apimnightlywebui24386.blob.core.windows.net/issues-6949-debug-mode-step-4/index.html)
_Notes_: The deployed app is linked to the management API of APIM master. (Same login and password as APIM master)

### CODE REVIEW

File: gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-debug/src/main/java/io/gravitee/gateway/debug/reactor/handler/context/AttributeHelper.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import java.util.Map;
+
+public abstract class AttributeHelper {
+
+    public static Map<String, Object> filterAndSerializeAttributes(Map<String, Object> attributes) {
+        if (attributes == null) {
+            return null;
+        }
+
+        // FIXME: context-path must be removed as it does not make sense in the debug context
```
Comment (ytvnr) [inline gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-debug/src/main/java/io/gravitee/gateway/debug/reactor/handler/context/AttributeHelper.java]: the message is not really clear imho

File: gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-api/src/main/java/io/gravitee/gateway/handlers/api/ApiContextHandlerFactory.java [hunk 2]
```diff
@@ -290,6 +286,10 @@ public AuthenticationHandlerSelector authenticationHandlerSelector(Authenticatio
         return new DefaultAuthenticationHandlerSelector(authenticationHandlerManager);
     }
 
+    protected ExecutionContextFactory executionContextFactory(ComponentProvider componentProvider) {
```
Comment (gaetanmaisse) [inline gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-api/src/main/java/io/gravitee/gateway/handlers/api/ApiContextHandlerFactory.java]: 🦉

File: gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-debug/src/main/java/io/gravitee/gateway/debug/reactor/handler/context/AttributeHelper.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        Map<String, String> filteredAttributes = new HashMap<>();
+        attributes
+            .keySet()
+            .stream()
+            .filter(key -> !attributesToExclude.contains(key))
+            .filter(key -> attributes.get(key) != null)
+            .filter(key -> attributes.get(key) instanceof Serializable)
+            .forEach(
+                key -> {
+                    filteredAttributes.put(key, attributes.get(key).toString());
```
Comment (gaetanmaisse) [inline gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-debug/src/main/java/io/gravitee/gateway/debug/reactor/handler/context/AttributeHelper.java]: Remove `toString`?

File: gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-debug/src/main/java/io/gravitee/gateway/debug/reactor/handler/context/AttributeHelper.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        if (attributes == null) {
+            return null;
+        }
+
+        // FIXME: context-path is removed for now as it is updated with the event id in the debug mode context
+        // we need to rework that before making it accessible to the user
+        // https://github.com/gravitee-io/issues/issues/7072
+        List<String> attributesToExclude = List.of("gravitee.attribute.context-path");
+
+        Map<String, String> filteredAttributes = new HashMap<>();
```
Comment (gaetanmaisse) [inline gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-debug/src/main/java/io/gravitee/gateway/debug/reactor/handler/context/AttributeHelper.java]:
```suggestion
        Map<String, Serializable> filteredAttributes = new HashMap<>();
```

File: gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-debug/src/main/java/io/gravitee/gateway/debug/reactor/processor/DebugEventCompletionProcessor.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * @author Yann TAVERNIER (yann.tavernier at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+public class DebugEventCompletionProcessor extends AbstractProcessor<ExecutionContext> {
```
Comment (gaetanmaisse) [inline gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-debug/src/main/java/io/gravitee/gateway/debug/reactor/processor/DebugEventCompletionProcessor.java:44]: DebugExecutionContext ?

File: gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-debug/src/test/java/io/gravitee/gateway/debug/reactor/handler/context/steps/DebugRequestStepTest.java [hunk 6]
```diff
@@ -195,63 +194,65 @@ public void shouldHaveAttributesInDiffMap() {
...
+        assertThat(resultDiff.containsKey(DebugStep.DIFF_KEY_BODY_BUFFER)).isFalse();
     }
 
     @Test
     public void shouldHaveBufferInDiffMap() {
         applyDifferentValuesForFields(List.of());
 
         cut.before(beforeRequest, new HashMap<>());
-        cut.after(afterRequest, new HashMap<>(), null, Buffer.buffer("outPut"));
+        cut.after(afterRequest, new HashMap<>(), Buffer.buffer("input"), Buffer.buffer("output"));
```
Comment (gaetanmaisse) [inline gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-debug/src/test/java/io/gravitee/gateway/debug/reactor/handler/context/steps/DebugRequestStepTest.java]: To remove

File: gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-debug/src/main/java/io/gravitee/gateway/debug/reactor/processor/DebugEventCompletionProcessor.java (hunk 5 above)
Comment (gaetanmaisse) [reply gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-debug/src/main/java/io/gravitee/gateway/debug/reactor/processor/DebugEventCompletionProcessor.java:44]: For bi variance issue we can't use `AbstractProcessor< DebugExecutionContext >`

**Review by ytvnr — APPROVED** [review]

**Pull request merged**